### ISSUE TITLE
Peers/Host data structures

### ISSUE THREAD

**Issue by kilrau** [ExchangeUnion/xud#185]
Redesign Peers/Host data structures to support future features:
- pubkey based on blocking #162 
- reputation
- get hosts packet response

**Comment by kilrau** [comment]
discussion with @sangaman

**Comment by moshababo** [comment]
Here's a rough design, subject to change. I'm posting it early so you could ask questions / comment about it before i'll revise it. We can edit this message for future updates.

**db tables**

* **hosts**: id (pk), score (def = 0, sum snapshot of reputation history), latestIpAddress, latestNodePubKey (nullable), latestListenPort (nullable)

* **hostIds**: hostId (fk), ipAddress (nullable), nodePubKey (nullable), createdAt

(each row assign either `ipAddress` **or** `nodePubKey` to a `hostId`. The purpose of this table is to associate multiple `ipAddress` and `nodePubKey` with a `hostId`. Reason: a recognizable host entity can have multiple ip addresses and multiple public keys. A new host entity will be instantiated based on `ipAddress`)

* **hostReputation**: id (pk), hostId (fk), event (enum), createdAt

(not specifying score of the event, for easy upgrades)

**app (pool)**

`peer` will be created together with a `host` object, based on `ipAddress`, which is known in the instantiation stage. `host` object will be the gateway to all updates by `peer`.

**create a new host (ipAddress, listenPort)**

1) insert into `hosts` (ipAddress, latestListenPort)
2) insert into `hostIds` (hostId, ipAddress)

**get host (ipAddress)**

select from `hostIds` (by ipAddress)

**xud postinstall**

create new hosts from a list of <ipAddress, listenPort>

 **p2p bootstrap:**

select top X from `hosts`, where `listenPort` is not null and `score` > Y, order by `score`

**new outbound**  

check if ipAddress is already connected (peerList), or if it's the node own ipAddress + listenPort. if not, get or create host, and instantiate `peer` with it. 

**new inbound** 

get or create host, and instantiate `peer` with it. if host score is below threshold, disconnect

**`get_hosts` request iteration**

for each host, new outbound connection

**`get_hosts` packet handling**

return select top X from `hosts`, where `listenPort` is not null and `score` > Y, order by `score`

**after outbound connection failure**

add negative reputation to host (new entry on `hostReputation` table, score is update on `hosts` table)

**after outbound connection success**

add positive reputation to host 

**after handshake**

- `nodePubKey` wasn't advertised or invalid - add negative reputation to host and disconnect from peer
- `nodePubKey` equals host's `latestNodePubKey` - do nothing
- `nodePubKey` not equals host's `latestNodePubKey` - it's either a new identity / the first time we're getting it, or a new `ipAddress` for an existing identity. if `nodePubKey` doesn't exists in `hostIds` table, then it's a new one, and we can just update `hosts` row, and insert it into `hostIds`. if it exists, verify that `latestIpAddress` is associated with its `hostId` (`ipAddress` in `hostIds` table). if so, just update `latestNodePubKey`. if not, update `ipAddress`'s `hostId` (`hostIds` table), and recreate/update the host object (because it's based on an incorrect PK)

**manual disconnection**

if using `ipAddress`, just get it from `peerList`, and destroy.
if using `nodePubKey`, iterate on `peerList`'s each peer host object, and look for the matching `nodePubKey`.

**Comment by sangaman** [comment]
For hosts, are we allowing a "host" to operate under more than one pubkey? My thinking is that pubkey is the primary identifier for a known peer or host, whose IP address might change but whose pubkey shouldn't. I can't think of a legitimate reason to change pubkey, at least. I'd think pubkey should be only **hosts**.

Also I don't think we should track the score on **hosts**. There's some possibility for the numerical score  and the sum of hostReputation events to disagree. I think it's fine to track the score as a number on a running process, but I'd prefer only persisting the events that change the score. Then when the process is restarted, it calculates the score on initialization.

**Comment by moshababo** [comment]
> For hosts, are we allowing a "host" to operate under more than one pubkey? My thinking is that pubkey is the primary identifier for a known peer or host, whose IP address might change but whose pubkey shouldn't. I can't think of a legitimate reason to change pubkey, at least.

Clients might change their pubkey due to a lost password or bad reputation. We can possibly link it to the same host identity, if they will reuse an IP address. We can decide not to do so, but still a host will need to be identified by IP address, because that's what we have when connection is established. And if we want to associate a host with multiple IP addresses, we'll still need a separate table to track them (many to one). 
We can't use pubkey as the pk (primary key) for `hosts`, because we're getting it only after the handshake (and if we associate a host with a peer only after handshake, we can't penalize it efficiently). So the pk will just be an incremented integer, and the main query identifier will be the list of IP addresses from a different table, which will be associated with it. We can keep track of `latestIpAddress` in `hosts`, but for now let's assume that we don't.
If an unknown IP address is being used (for inbound/outbound connection), a new host identity will be created. If a known pubkey will be advertised later, it will get coalesced with the previous one, and the new IP address will be associated with that pubkey.
If a known IP address is being used, but will later advertise an different pubkey from previously - we'll create a new host identity for it, and remove the association of that IP address with the previous pubkey.

Let me know if that was what you meant, and if you want to proceed with this approach.

> Also I don't think we should track the score on hosts. There's some possibility for the numerical score and the sum of hostReputation events to disagree. I think it's fine to track the score as a number on a running process, but I'd prefer only persisting the events that change the score. Then when the process is restarted, it calculates the score on initialization.

It's definitely better to calculated it from the list of events (because if not, it will certainly get out-of-sync at some point). I thought to add it as a snapshot to reduce the hassle of recalculating it on every request. But since we can cache the ongoing score, we can drop that. (it can also get out-of-sync, but it's not a big deal)

**Comment by kilrau** [comment]
Quick notes, to be revised:

one identifier
one pubkey - multiple ip addresses
pre-handshake spam
hostid local random pre-handshake

**Comment by moshababo** [comment]
> (No second hostIds table since 1) a new nodePubKey = hostId and we only have hostId because we get the nodePubKey after a successful handshake and 2) IPs change too frequently to be worthy to save them. When we connect to a new ipAddress, we instantiate a new host object and if it turns out we know the nodePubKey we just update the IP of the existing hostID and delete the newly created hostID. Added stake check timestamp.

So you're saying that we should keep only one IP address per `hostId`/`nodePubKey`?

> TBD: Before you marked listenPort as nullable. How would we connect to a host without knowing the port? I mentioned before that we want to be as censorship resistant as possible. How about getting the actual listenPort via an https request on 443? @moshababo @sangaman)

For an inbound connection, the port that we know of is meaningless (was just randomly generated by the peer's OS), so we won't have his `listenPort` until he will advertise it (same like `nodePubKey`).

How will an HTTPS request on 443 will help here?

> app (pool) (Why this naming? @moshababo)

There was no renaming here. I just specified how `pool` will connect between `peer` and `host` objects.

> create new hosts from a list of <ipAddress, listenPort>
(please add details @moshababo)

We will maintain an hardcoded seed list of hosts (<ipAddress, listenPort>) in our repo. after installing `xud` dependencies, we will initialize them to the database, where that list will continue to grow dynamically (from manually connecting to new hosts, or from `get_hosts` iterations to seed hosts).

> select top X from hosts, where nodePubKey & listenPort is not null and score > Y, order by score
(Y default is -1)

Unlike what we previously implemented, we plan to create new host entities before we ever established a successful connection with them. So I would say that Y should be the score that represent a host after a successful connection + handshake.

**Comment by moshababo** [comment]
After we'll agree on the basic general approach in regards to IP address <-> pubKey relationship (many to one / one to one / zero to one), i'll update my original message and make it more clear.

**Comment by kilrau** [comment]
Commented directly above. I am in favor of a one to one IP address <-> pubKey relationship. Let's see what @sangaman says.

**Comment by sangaman** [comment]
I think I prefer allowing more than 1 IP address per pubkey, although I agree we shouldn't need to track old IP addresses since they can indeed change frequently and the old ones become obsolete. LND allows a node to advertise multiple connection URIs. Maybe we can follow the behavior Kilian described, tracking only 1 and overwriting the current IP address value if we find the same pubkey on a different IP, but also allow nodes to share multiple IP addresses during the handshake that we can use for reconnection attempts.

Also for failed connection attempts, I think a backoff approach might make sense. Try to connect to known hosts after 1 minute, 2 minutes, 4 minutes, etc... until success. We can maybe set some hard upper limit at which point we remove their IP address from the table so that we don't try connecting to them again.

**Comment by moshababo** [comment]
@kilrau I replied again on your comments above. 

--

> I think I prefer allowing more than 1 IP address per pubkey, although I agree we shouldn't need to track old IP addresses since they can indeed change frequently and the old ones become obsolete. LND allows a node to advertise multiple connection URIs. Maybe we can follow the behavior Kilian described, tracking only 1 and overwriting the current IP address value if we find the same pubkey on a different IP, but also allow nodes to share multiple IP addresses during the handshake that we can use for reconnection attempts.

Tracking old IP addresses can help us to immediately associate a connection with a host/nodePubKey, mostly for blocking purposes, not for reconnection attempts. This can become out of hand if an exchange will get someone else's old IP address.
We can instead allow peers to control the IP addresses they are associated with by advertising them, and that will be used solely for reconnection attempts.  

> Also for failed connection attempts, I think a backoff approach might make sense. Try to connect to known hosts after 1 minute, 2 minutes, 4 minutes, etc... until success. We can maybe set some hard upper limit at which point we remove their IP address from the table so that we don't try connecting to them again.

Sounds reasonable, are you suggesting that instead of penalizing, so that only the upper limit would be the point in which we penalize?

**Comment by sangaman** [comment]
> Tracking old IP addresses can help us to immediately associate a connection with a host/nodePubKey, mostly for blocking purposes, not for reconnection attempts.

I think I'd prefer only blocking IPs that have actively misbehaved against us, it feels excessive to ban every IP a misbehaving peer has ever used. Either way though it doesn't seem to be that big of a deal.

> We can instead allow peers to control the IP addresses they are associated with by advertising them, and that will be used solely for reconnection attempts.

Agreed.

> Sounds reasonable, are you suggesting that instead of penalizing, so that only the upper limit would be the point in which we penalize?

Yeah, we're just talking about attempts to reconnect to known peers right? I wouldn't want to penalize peers simply for being offline for a while. And even when they hit the upper limit, I don't think I'd punish necessarily but just forget about their IP address and don't retry outgoing connections to them.

**Comment by sangaman** [comment]
I'd be interested in working on this once we finalize the database structure.

**Comment by moshababo** [comment]
@sangaman no probs, after we'll finalize the design we need to go through on all the different applications of this, and open the appropriate issue/s.

**Comment by moshababo** [comment]
> Maybe we should track banned IP addresses in a separate table along with banned hosts. This way we don't have to "forget" IP addresses we had previously banned. The vulnerability I can think about here is that a misbehaving peer could alternate IP addresses and endlessly attempt handshakes as a sort of DOS attack.

If we continue to track them in a separate table, and don't block them, what's the purpose of that list? And if we block them, how this differ from what I originally suggested? If it's just that you want to apply blocking to specific IP address, and not to the entire history - I don't see how it can help, while it add more complexity. We want to block based on `hostId`/`nodePubKey`, and whatever IP addresses associated with it (not including reconnect-attempts-only addresses, which the peer's control).

For the vulnerability you described, if the attacked will continuously use new IP addresses, there's nothing much we can do, except disconnect from him after a bad handshake. 

To summerize: if we block the entire IP addresses history, we might block other exchanges that ended up with the host previous IP address (i'm not sure how common this would be). If we block only the latest IP address or no IP address, we are **more** vulnerable to DoS attacks. 

> Alternatively, maybe we track the last time we were connected to a peer, and we don't attempt to connect to them if it's been too long. I don't want to "block" the host because I don't want to stop them from creating an inbound connection to us should they come back online at some point. But I also don't want to endlessly continue attempting reconnections to "zombie" hosts that have permanently gone offline.

Blocking would require much higher negative reputation then dropping it from the list for the p2p bootstrap process. The main question now is whether we should use the same reputation mechanism for both, by setting different thresholds for each, or should we maintain a separate mechanism.

If we don't connect with a peer (p2p bootstrap process) solely because we weren't connected with it recently, it can also happen if he was continuously available, but `xud` wasn't activated. We can instead maintain `lastConnectionFailure` for each, and play with this one, along with reconnection interval and penalization. I can offer a solution for this one on a separate issue. We just need to finalize first on whether it should be associated with the reputation or not. I don't see a reason why not.

**Comment by kilrau** [comment]
> Tracking old IP addresses can help us to immediately associate a connection with a host/nodePubKey, mostly for blocking purposes, not for reconnection attempts. This can become out of hand if an exchange will get someone else's old IP address.
> To summerize: if we block the entire IP addresses history, we might block other exchanges that ended up with the host previous IP address (i'm not sure how common this would be). If we block only the latest IP address or no IP address, we are more vulnerable to DoS attacks.

We still shouldn't do that. Getting new IP addresses is easy and cheap, I'm pretty convinced we won't solve DDoS on IP-level, that's why companies like CloudFlare exist :). 

> For the vulnerability you described, if the attacker will continuously use new IP addresses, there's nothing much we can do, except disconnect from him after a bad handshake.

Exactly. We might find a way to punish via staked coins in a later stage. Suma Sumarum I suggest doing blocking only based on `nodePubKey` for now.

> We can instead allow peers to control the IP addresses they are associated with by advertising them, and that will be used solely for reconnection attempts.

Agreed. So second table for IP addresses, but only advertised by the node itself (signed) and only for reconnection attempts which are triggered by me.

> I don't want to "block" the host because I don't want to stop them from creating an inbound connection to us should they come back online at some point. But I also don't want to endlessly continue attempting reconnections to "zombie" hosts that have permanently gone offline.

Also agree on this one, let's research how others do it.

Blocking would require much higher negative reputation then dropping it from the list for the p2p bootstrap process. The main question now is whether we should use the same reputation mechanism for both, by setting different thresholds for each, or should we maintain a separate mechanism.

> If we don't connect with a peer (p2p bootstrap process) solely because we weren't connected with it recently, it can also happen if he was continuously available, but xud wasn't activated. We can instead maintain lastConnectionFailure for each, and play with this one, along with reconnection interval and penalization. I can offer a solution for this one on a separate issue. We just need to finalize first on whether it should be associated with the reputation or not. I don't see a reason why not.

Even though reputation includes more than only successful/failed connection attempts it's a good indicator which XUDs should be onthe bootstrap list IMO. So yes, why not.

**Comment by sangaman** [comment]
> If we continue to track them in a separate table, and don't block them, what's the purpose of that list? And if we block them, how this differ from what I originally suggested? If it's just that you want to apply blocking to specific IP address, and not to the entire history - I don't see how it can help, while it add more complexity. We want to block based on hostId/nodePubKey, and whatever IP addresses associated with it (not including reconnect-attempts-only addresses, which the peer's control).
> 
> For the vulnerability you described, if the attacker will continuously use new IP addresses, there's nothing much we can do, except disconnect from him after a bad handshake.
> 
> To summerize: if we block the entire IP addresses history, we might block other exchanges that ended up with the host previous IP address (i'm not sure how common this would be). If we block only the latest IP address or no IP address, we are more vulnerable to DoS attacks.

> We still shouldn't do that. Getting new IP addresses is easy and cheap, I'm pretty convinced we won't solve DDoS on IP-level, that's why companies like CloudFlare exist...

I see what you mean. After thinking about this some more, I'm questioning the value of persisting IP address bans at all - given the complexity involved and how easy it can be to get hew IP addresses. Instead, maybe we should just maintain an IP ban list for the inside a running instance but not save bans to the database for the long term. And we'd only ban IP addresses for failing handshakes. 

> Blocking would require much higher negative reputation then dropping it from the list for the p2p bootstrap process. The main question now is whether we should use the same reputation mechanism for both, by setting different thresholds for each, or should we maintain a separate mechanism.

I'd prefer a separate mechanism.

> If we don't connect with a peer (p2p bootstrap process) solely because we weren't connected with it recently, it can also happen if he was continuously available, but xud wasn't activated. We can instead maintain lastConnectionFailure for each, and play with this one, along with reconnection interval and penalization. I can offer a solution for this one on a separate issue. We just need to finalize first on whether it should be associated with the reputation or not. I don't see a reason why not.

Yeah that sounds good, we can discuss that approach separately. I'd prefer it to be separate from reputation because I consider negative reputation to be misbehavior or unreliability, and simply being offline seems like neither of those two things.

**Comment by kilrau** [comment]
Moving this to implementation in alpha

**Comment by kilrau** [comment]
- temp IP ban:

**Comment by sangaman** [comment]
tempBannedIPs shouldn't be a table, I don't think we'll need to persist it between sessions.

Also, I think that rather than **bannedHosts** we should just add a `banned` column to the `hosts` table and maybe a `bannedAt` timestamp column.

**Comment by kilrau** [comment]
> **tempBannedIPs** shouldn't be a table, I don't think we'll need to persist it between sessions.

Also, I think that rather than **bannedHosts** we should just add a `banned` column to the `hosts` table and maybe a `bannedAt` timestamp column.

You are right, didn't think too much obviously. Fixed.

**Comment by kilrau** [comment]
Do we already have any sort of logic on how and how often we try to reconnect to hosts?

**Comment by kilrau** [comment]
Putting this into alpha.2, since https://github.com/ExchangeUnion/xud/pull/315 doesn't close this completely and Daniel's focus will be the swap for alpha.2

**Comment by kilrau** [comment]
this issue is fairly outdated, can you help me to rephrase what we need to restructure to support
- pubkey based on blocking #162 
- reputation #151 

? @moshababo @sangaman

**Comment by sangaman** [comment]
Yes. Perhaps we should close this issue though and create separate issues for reputation, banning/ban scores, etc... I can do this if you agree.

**Comment by kilrau** [comment]
Agree! @sangaman
